File: UVa/10063/10063.py
from sys import stdin

sols = []
l = str()

def solve(sol, idx, it):
	global sols, l
	if len(sol) == len(l):
		sols.append(sol)
	else:
		index = 0
		for i in range(idx, idx + it):
			sc = sol.copy()
			sc.insert(index, idx)
			solve(sc, idx + 1, it + 1)
			index += 1

def main():
	global sols, l
	l = stdin.readline().strip()
	cases = 0

	while(l != ""):
		if(cases != 0):
			print()
		sols = []
		solve([0], 1, 2)

		for i in range(len(sols)):
			sc = ""
			for j in range(len(sols[i])):
				sc += l[(sols[i][j])]
			print(sc)
		cases += 1

		l = stdin.readline().strip()

main()

File: UVa/10063/10063.cpp
#include <bits/stdc++.h>

using namespace std;

string str;
vector <vector <int>> sols;
int n;

void solve(vector <int> sol, int idx, int it){
	if(sol.size() == n)
		sols.push_back(sol);
	else{
		int index = 0, i = idx;
		vector <int> sc;
		for(; i < it + idx; i++){
			sc = sol;
			sc.insert(sc.begin() + index++, idx);
			solve(sc, idx + 1, it + 1);
		}
	}
}

int main(){
	int cases = 0;
	while(cin >> str){
		if(cases++)
			printf("\n");
		sols = vector <vector <int>>();
		n = str.size();
		solve({0}, 1, 2);
		string temp;

		for(int i = 0; i < sols.size(); i++){
			temp = "";
			for(int j = 0; j < n; j++)
				temp.push_back(str[sols[i][j]]);
			cout << temp << endl;
		}
	}

	return 0;
}